### PR TITLE
Link System.Globalization.Native into coreclr

### PR DESCRIPTION
**Pull request by safern**
Contributes to: https://github.com/dotnet/runtime/issues/826

This is the first effort of using ICU on windows. Linking S.G.Native into coreclr.

Next step is to add qcalls and change all the calls into S.G.Native to be runtime calls.

Thanks @jkoritzinsky for your help on plumbing this through 😄

### CODE REVIEW

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt [hunk 1]
```diff
@@ -92,11 +92,21 @@ set_target_properties(System.Globalization.Native-Static PROPERTIES OUTPUT_NAME
...
 
 if(NOT CMAKE_SYSTEM_NAME STREQUAL Darwin)
-    add_custom_command(TARGET System.Globalization.Native POST_BUILD
-        COMMENT "Verifying System.Globalization.Native.so dependencies"
-        COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/../verify-so.sh
-            $<TARGET_FILE:System.Globalization.Native>
-            "Verification failed. System.Globalization.Native.so has undefined dependencies. These are likely ICU APIs that need to be added to icushim.h."
-        VERBATIM
-    )
+    if (BUILD_GLOBALIZATION_SHARED_LIBRARY)
```
Comment (jkotas) [inline src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: Other existing shim CMakefiles use `GEN_SHARED_LIB` for this purpose. Do the same for consistency?

https://github.com/dotnet/runtime/search?utf8=%E2%9C%93&q=GEN_SHARED_LIB&type=

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt [hunk 2]
```diff
@@ -57,24 +57,32 @@ set(NATIVEGLOBALIZATION_SOURCES
...
+endif()
+
 include_directories(${UTYPES_H})
 
-add_library(System.Globalization.Native
-    SHARED
-    ${NATIVEGLOBALIZATION_SOURCES}
-    ${VERSION_FILE_PATH}
-)
+if (BUILD_GLOBALIZATION_SHARED_LIBRARY)
```
Comment (jkotas) [inline src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: Other existing shim CMakefiles use `GEN_SHARED_LIB` for this purpose. Do the same for consistency?

https://github.com/dotnet/runtime/search?utf8=%E2%9C%93&q=GEN_SHARED_LIB&type=

File: src/coreclr/src/librariesshims/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+set(EXCLUDE_VERSION_HEADER_FILE ON)
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+
+if(CLR_CMAKE_HOST_UNIX AND NOT DEFINED CLR_CROSS_COMPONENTS_BUILD)
+    set(BUILD_GLOBALIZATION_SHARED_LIBRARY OFF)
+    include_directories("../../../libraries/Native/Unix/Common")
```
Comment (jkotas) [inline src/coreclr/src/librariesshims/CMakeLists.txt]: Why do you need this include here? Would it be better for `System.Globalization.Native` to have it inside?

File: src/coreclr/CMakeLists.txt [hunk 4]
```diff
@@ -129,6 +129,8 @@ endif(CLR_CROSS_COMPONENTS_BUILD)
 #-------------------
 include(pgosupport.cmake)
 
+add_subdirectory(src/librariesshims)
```
Comment (jkotas) [inline src/coreclr/CMakeLists.txt]: Nit: `librariesshims` is hard to parse name and associate with `libraries/native`. What about `libraries-native` ?

File: src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt [hunk 5]
```diff
@@ -123,6 +123,7 @@ else()
         ${END_WHOLE_ARCHIVE}
         mscorrc_debug
         palrt
+        System.Globalization.Native-Static
```
Comment (jkotas) [inline src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt]: Just something to double check - is this binplaced a different directory than the one built from libraries/native? We do not want libraries build to be overwriting products of coreclr build.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt (hunk 2 above)
Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: Good catch, will update.

File: src/coreclr/src/librariesshims/CMakeLists.txt (hunk 3 above)
Comment (safern) [reply src/coreclr/src/librariesshims/CMakeLists.txt]: Because `src/libraries/Native/Unix` adds that include and we don't include that file on this path. However, I could set a variable and then condition the include in `System.Globalization.Native` when that variable is ON

File: src/coreclr/CMakeLists.txt (hunk 4 above)
Comment (safern) [reply src/coreclr/CMakeLists.txt]: Good idea, will update.

File: src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt (hunk 5 above)
Comment (safern) [reply src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt]: Yeah, they will be binplaced in a different directory because when building coreclr, the output directory will be set by the coreclr cmakelists which points to coreclr output directory.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt [hunk 6]
```diff
@@ -57,24 +57,32 @@ set(NATIVEGLOBALIZATION_SOURCES
     pal_icushim.c
 )
 
+if (NOT EXCLUDE_VERSION_HEADER_FILE)
+    list(APPEND NATIVEGLOBALIZATION_SOURCES
```
Comment (jkotas) [inline src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: This should be excluded for all static builds. I do not think we need `EXCLUDE_VERSION_HEADER_FILE`.

Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: You mean, `NO GEN_SHARED_LIB`?

Comment (jkotas) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: I mean that it should be in the `add_library(System.Globalization.Native ...` target as it is today, and deleted from the `add_library(System.Globalization.Native-Static` target.

Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: So `System.Globalization.Native-Static` shouldn’t carry the version information?

Comment (jkotas) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: I do not see a good reason for it.

Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: Ok then I’ll change it to that.

**Comment by am11** [comment]
+1, do the sources compile on windows? last time when I attempted using icu on windows (when sources were still in dotnet/coreclr repo), there were some missing headers (`icucommon.h` and `icui18n.h`).

**Comment by safern** [comment]
> +1, do the sources compile on windows? 

Thanks for the input, @am11. Once I get to the point where we are doing runtime calls (QCalls) instead of regular PInvokes, I will start working on getting it building for windows.

File: src/coreclr/CMakeLists.txt [hunk 7]
```diff
@@ -129,6 +129,8 @@ endif(CLR_CROSS_COMPONENTS_BUILD)
 #-------------------
 include(pgosupport.cmake)
 
+add_subdirectory(src/libraries-native)
```
Comment (am11) [inline src/coreclr/CMakeLists.txt]: We can define `${CLR_REPO_ROOT_DIR}` just like we are using `${CLR_ENG_NATIVE_DIR}` on line 118, it is passed from:

https://github.com/dotnet/runtime/blob/932baca94e88ac77a5a70b2032d3341603b4610e/eng/native/build-commons.sh#L114-L117

```diff
- __CMakeArgs="-DCLR_ENG_NATIVE_DIR=\"$engNativeDir\" $__CMakeArgs" 
+ __CMakeArgs="-DCLR_ENG_NATIVE_DIR=\"$engNativeDir\" $__CMakeArgs -DCLR_REPO_ROOT_DIR=\"$__RepoRootDir\" $__CMakeArgs"
```

same thing at: https://github.com/dotnet/runtime/blob/932baca94e88ac77a5a70b2032d3341603b4610e/src/coreclr/build.cmd#L460

after that we can replace this line with:

```cmake
set(CMAKE_INCLUDE_CURRENT_DIR ON)
set(CORECLR_BUILD ON)

if(CLR_CMAKE_HOST_UNIX AND NOT DEFINED CLR_CROSS_COMPONENTS_BUILD)
    add_subdirectory(${CLR_REPO_ROOT_DIR}/src/libraries/Native/Unix/System.Globalization.Native System.Globalization.Native)
endif()
```

without a separate `src/coreclr/src/libraries-native/CMakeLists.txt` proxy file.

Comment (jkoritzinsky) [reply src/coreclr/CMakeLists.txt]: I’d still rather have it in a separate CMakeLists file because we don’t want to leak the CMAKE_INCLUDE_CURRENT_DIR setting or the other include directories.

**Comment by am11** [comment]
> Windows 19H1 should have ICU and the necessary headers from my understanding.

I could only find `icu.h` in %programfiles% and %programfiles(x86)% directories; on insider build `10.0.19546.1000`. Perhaps those headers weren't deemed important for Windows port or they get just installed differently. Globalization project is using ICU-C headers (not the C++ ones) - some Linux distros differentiate between the two (i.e. icuc is a separate sub-package).

File: src/coreclr/CMakeLists.txt (hunk 7 above)
Comment (am11) [reply src/coreclr/CMakeLists.txt]: one pattern which is used in certain places for that purpose is:

```cmake
set(SOMETHING ON)

# do stuff

set(SOMETHING) # without args, to unset
```

Comment (safern) [reply src/coreclr/CMakeLists.txt]: The proxy file was intended because in the future we might link more of the shims in order to be able to do runtime calls, so then we would just include and add subdirectories here without necessarily complicating `src/coreclr/CMakeLists.txt` -- also that way we keep the set and defines isolated for libraries-native, I like the separation as it is cleaner.

> We can define ${CLR_REPO_ROOT_DIR}

I thought about that, but talking to @jkoritzinsky he recommended to avoid passing more defines via command line as it is nice to keep the cmake command simple so that you basically can just do cmake and it builds fine without having to learn all the parameters.

Comment (am11) [reply src/coreclr/CMakeLists.txt]: @janvorli recommended to add `CLR_REPO_ROOT_DIR` instead of hardcoding relative paths in my previous PR. There are about dozen of arguments passed to the final cmake call (and counting), so i guess it would be aligned with where we are heading.

**Comment by tarekgh** [comment]
CC @jefgen who may help with the comment https://github.com/dotnet/runtime/pull/6928#issuecomment-580618896

File: src/coreclr/src/libraries-native/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,6 @@
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+set(CORECLR_BUILD ON)
+
+if(CLR_CMAKE_HOST_UNIX AND NOT DEFINED CLR_CROSS_COMPONENTS_BUILD)
+    add_subdirectory(../../../libraries/Native/Unix/System.Globalization.Native System.Globalization.Native)
```
Comment (tarekgh) [inline src/coreclr/src/libraries-native/CMakeLists.txt]: >../../../l [](start = 21, length = 10)

We don't have the repo root path stored in any variable instead of using relative path like that?

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt [hunk 9]
```diff
@@ -84,11 +89,21 @@ set_target_properties(System.Globalization.Native-Static PROPERTIES OUTPUT_NAME
...
+        add_custom_command(TARGET System.Globalization.Native POST_BUILD
+            COMMENT "Verifying System.Globalization.Native.so dependencies"
+            COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/../verify-so.sh
+                $<TARGET_FILE:System.Globalization.Native>
+                "Verification failed. System.Globalization.Native.so has undefined dependencies. These are likely ICU APIs that need to be added to icushim.h."
+            VERBATIM
+        )
+    else()
+        add_custom_command(TARGET System.Globalization.Native-Static POST_BUILD
+            COMMENT "Verifying System.Globalization.Native-Static.so dependencies"
```
Comment (tarekgh) [inline src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: >.so [](start = 65, length = 3)

just checking, does the static library built with .so extension?

File: src/coreclr/src/libraries-native/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,6 @@
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+set(CORECLR_BUILD ON)
+
+if(CLR_CMAKE_HOST_UNIX AND NOT DEFINED CLR_CROSS_COMPONENTS_BUILD)
```
Comment (janvorli) [inline src/coreclr/src/libraries-native/CMakeLists.txt]: Please remove the DEFINED here. It is not necessary. We define the variable to 1 (when the cross components build is on) and when it is not defined, it behaves as 0 or False in a condition.
I can see that we use both forms in the cmake files and I would like to standardize on the simpler form.

File: src/libraries/Native/Unix/CMakeLists.txt [hunk 11]
```diff
@@ -296,6 +296,9 @@ endif()
 add_subdirectory(System.Native)
 
 if (NOT CLR_CMAKE_PLATFORM_WASM)
+    # In coreclr we currently link to it as part of the coreclr build.
+    # For libraries and mono build we need to the shared library until mono links to it.
```
Comment (janvorli) [inline src/libraries/Native/Unix/CMakeLists.txt]:
```suggestion
    # For libraries and mono build we need to build the shared library until mono links to it.
```

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt [hunk 12]
```diff
@@ -59,22 +59,27 @@ set(NATIVEGLOBALIZATION_SOURCES
 
 include_directories(${UTYPES_H})
 
-add_library(System.Globalization.Native
-    SHARED
-    ${NATIVEGLOBALIZATION_SOURCES}
-    ${VERSION_FILE_PATH}
-)
+if (CORECLR_BUILD)
+ include_directories("../Common")
```
Comment (janvorli) [inline src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: Can you please fix the indentation here to 4 spaces?

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt [hunk 13]
```diff
@@ -59,22 +59,27 @@ set(NATIVEGLOBALIZATION_SOURCES
...
 
-install_library_and_symbols (System.Globalization.Native)
+    target_link_libraries(System.Globalization.Native
+        dl
+    )
+
+    install_library_and_symbols (System.Globalization.Native)
+endif()
 
 add_library(System.Globalization.Native-Static
```
Comment (janvorli) [inline src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: I wonder why do we generate the static library when building libraries. Was it ever used by anything? If not, then I'd prefer naming the static library the same as the dynamic one. The coreclr build is composed of many static libraries and none of them have a suffix indicating it is static.

File: src/coreclr/CMakeLists.txt (hunk 7 above)
Comment (janvorli) [reply src/coreclr/CMakeLists.txt]: I would also prefer not to use the long relative paths that are hard to read. Defining `CLR_REPO_ROOT_DIR` would be my preference. As @am11 says, we already have quite some defines that are necessary.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt [hunk 14]
```diff
@@ -84,11 +89,21 @@ set_target_properties(System.Globalization.Native-Static PROPERTIES OUTPUT_NAME
...
+    if (GEN_SHARED_LIB)
+        add_custom_command(TARGET System.Globalization.Native POST_BUILD
+            COMMENT "Verifying System.Globalization.Native.so dependencies"
+            COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/../verify-so.sh
+                $<TARGET_FILE:System.Globalization.Native>
+                "Verification failed. System.Globalization.Native.so has undefined dependencies. These are likely ICU APIs that need to be added to icushim.h."
+            VERBATIM
+        )
+    else()
+        add_custom_command(TARGET System.Globalization.Native-Static POST_BUILD
```
Comment (janvorli) [inline src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: It doesn't make sense to verify the static library. It would not work anyways. The verification is meant only for shared libraries.

File: src/coreclr/src/libraries-native/CMakeLists.txt (hunk 8 above)
Comment (janvorli) [inline src/coreclr/src/libraries-native/CMakeLists.txt]: Is the binary_dir argument necessary here? We don't use such a convention anywhere

File: src/coreclr/src/libraries-native/CMakeLists.txt (hunk 10 above)
Comment (safern) [reply src/coreclr/src/libraries-native/CMakeLists.txt]: Sounds good. I'll remove it, thanks.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt (hunk 14 above)
Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: Ok, I'll remove it then.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt (hunk 13 above)
Comment (jkotas) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: The static builds of the shims are used by CoreRT and Mono/WebAssembly. Mono/Webassembly does not use System.Globalization currently, but it does use some of the other ones.

> suffix indicating it is static.

IIRC, this suffix was added to make cmake(?) happy.

File: src/coreclr/src/libraries-native/CMakeLists.txt (hunk 8 above)
Comment (safern) [reply src/coreclr/src/libraries-native/CMakeLists.txt]: If we don't specify it, then it would be under `libraries-native` in the intermediate directory instead?

File: src/coreclr/CMakeLists.txt (hunk 7 above)
Comment (safern) [reply src/coreclr/CMakeLists.txt]: Ok sounds good, I'll add that as well.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt (hunk 9 above)
Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: No, I forgot to remove it, it builds with `.a`

File: src/coreclr/src/libraries-native/CMakeLists.txt (hunk 8 above)
Comment (safern) [reply src/coreclr/src/libraries-native/CMakeLists.txt]: Nope, but I'm going to add it.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt (hunk 13 above)
Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: I'm not sure if it was used before, we do that for all the native libraries we build in libraries. I believe for wasm we need to build them, but when building wasm, we don't build this library nor the shared libraries. I can update this to only build when `NOT GEN_SHARED_LIB`.

File: src/libraries/Native/Unix/CMakeLists.txt (hunk 11 above)
Comment (safern) [reply src/libraries/Native/Unix/CMakeLists.txt]: I can remove this as we changed to use `GEN_SHARED_LIB` which is set when not building in WASM for libraries.

File: src/coreclr/src/libraries-native/CMakeLists.txt (hunk 8 above)
Comment (safern) [reply src/coreclr/src/libraries-native/CMakeLists.txt]: We need to specify a directory because the subdirectory that we're adding is not a subdirectory of src/coreclr/src/libraries-native.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt (hunk 13 above)
Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: I'll leave as is given that we might still need it for Mono/WebAssembly in the future. When mono links to this library we can stop building the SHARED library at all and then remove the -Static suffix.

File: src/coreclr/src/libraries-native/CMakeLists.txt (hunk 8 above)
Comment (safern) [reply src/coreclr/src/libraries-native/CMakeLists.txt]: I left it as is as without it it breaks, any suggestion, I would appreciate it.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt (hunk 13 above)
Comment (janvorli) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: The `-Static` suffix (or any other way to differentiate the static and dynamic libraries) is needed when both dynamic and static versions are produced. If we were producing just one, it would not be needed. I was referring to the case when we would be building just one flavor.

Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: Yeah, I understand that. When building libraries we build both as of now. Do we want to stop building the Static as part of libraries build?

Comment (janvorli) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: Based on the discussion above, let's keep it.

File: src/libraries/Native/Unix/CMakeLists.txt [hunk 15]
```diff
@@ -285,20 +285,14 @@ include(${CLR_ENG_NATIVE_DIR}/configuretools.cmake)
...
     add_compile_options(-Weverything)
 endif()
 
 add_subdirectory(System.Native)
 
 if (NOT CLR_CMAKE_PLATFORM_WASM)
-    # In coreclr we currently link to it as part of the coreclr build.
-    # For libraries and mono build we need to the shared library until mono links to it.
-    set(BUILD_GLOBALIZATION_SHARED_LIBRARY ON)
+	add_subdirectory(System.IO.Ports.Native)
```
Comment (janvorli) [inline src/libraries/Native/Unix/CMakeLists.txt]: A nit - wrong indentation

File: src/coreclr/src/libraries-native/CMakeLists.txt (hunk 8 above)
Comment (janvorli) [reply src/coreclr/src/libraries-native/CMakeLists.txt]: I see, cmake complains about it. The CMake doc was a bit unclear on this, it seemed to me as if the non-relative part of the src path would be used relative to the current bin dir, but that's obviously not the case.

File: src/libraries/Native/Unix/CMakeLists.txt (hunk 15 above)
Comment (safern) [reply src/libraries/Native/Unix/CMakeLists.txt]: 😢 good catch!

**Comment by jefgen** [comment]
> > Windows 19H1 should have ICU and the necessary headers from my understanding.
> > I could only find `icu.h`.

Note: If you're limiting yourself to Windows 19H1 and above, then you'll want to use the header `icu.h` and link to `icu.lib` (which is for `icu.dll`) instead.

There is some history behind the various headers... but the short version is: Just use `icu.h` and `icu.lib`/`icu.dll` going forward.

The super-short history is that upstream ICU made a binary breaking change in version 63.1, which necessitated the creation of mini shim layer over top of the previous binaries (`icuuc.dll` and `icuin.dll`) This new binary (`icu.dll`) contains both, and provides the stable C API surface for `icu.h`.

**Comment by safern** [comment]
Build is green and changes are ready 🎉

**Comment by am11** [comment]
@jefgen, thanks for summarizing. Are the [`unicode/*`](https://github.com/unicode-org/icu/tree/master/icu4c/source/common/unicode) headers also available somewhere, I couldn't find these on my box: https://github.com/dotnet/runtime/blob/6c8a6290cb150d903e514322c1637366371a5742/src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.h#L16-L35

**Comment by jefgen** [comment]
> Are the [`unicode/*`](https://github.com/unicode-org/icu/tree/master/icu4c/source/common/unicode) headers also available somewhere?

Thanks for the question! The various `unicode/*` headers are intentionally *not* available anywhere in the Windows SDK, as they contain both C and C++ APIs, as well as draft, unstable, and internal ICU APIs which we cannot expose in the SDK.

The `icu.h` header file is generated from the `unicode/*` headers by filtering out the C++/draft/unstable/internal APIs, and then all of the headers are merged/concatenated together into a single file.

As an aside: FWIW, you could sort of approximate the same thing with the raw/real ICU headers by defining the following values before including any headers:

```c
#define U_SHOW_CPLUSPLUS_API 0
#define U_DEFAULT_SHOW_DRAFT 0
#define U_HIDE_DRAFT_API 1
#define U_HIDE_DEPRECATED_API 1
#define U_HIDE_OBSOLETE_API 1
#define U_HIDE_INTERNAL_API 1
```

**Comment by mjsabby** [comment]
Will this feature be 19H1+ only?

**Comment by mjsabby** [comment]
> Will this feature be 19H1+ only?

Answered my own question. From the linked issue @tarekgh mentions it will be 19H1+.

**Comment by jkotas** [comment]
And you are going to have an option to bundle the ICU with your app and use that everywhere.

**Comment by safern** [comment]
@jkotas @janvorli is this good to merge?

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt [hunk 16]
```diff
@@ -59,22 +59,27 @@ set(NATIVEGLOBALIZATION_SOURCES
...
+        dl
+    )
+
+    install_library_and_symbols (System.Globalization.Native)
+endif()
 
 add_library(System.Globalization.Native-Static
     STATIC
     ${NATIVEGLOBALIZATION_SOURCES}
-    ${VERSION_FILE_PATH}
```
Comment (jkotas) [inline src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: It would be nice to delete it from all static libs for consistency.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt [hunk 17]
```diff
@@ -59,22 +59,27 @@ set(NATIVEGLOBALIZATION_SOURCES
 
 include_directories(${UTYPES_H})
 
-add_library(System.Globalization.Native
-    SHARED
-    ${NATIVEGLOBALIZATION_SOURCES}
-    ${VERSION_FILE_PATH}
-)
+if (CORECLR_BUILD)
```
Comment (jkotas) [inline src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: I would call this something else to future proof this for when Mono is going to need to do the same.

Or just not have this at all, and add `include_directories("../Common")` to every shim unconditionally.

Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: I could get away by just including it here unconditionally and leaving it for other shims in the common CMakeLists in the `Unix` root directory.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt (hunk 16 above)
Comment (safern) [reply src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt]: Also fixed.

**Review by janvorli — APPROVED** [review]

**Review by jkotas — APPROVED** [review]

**Comment by safern** [comment]
Test failure is: https://github.com/dotnet/runtime/issues/30732

**Comment by safern** [comment]
Thanks for all your reviews 😄
